OpenSim/Simulation/Model/Appearance.h

```diff
@@ -33,105 +33,157 @@ namespace OpenSim {
 class Model;
 
-
-//=============================================================================
-//=============================================================================
+using SimTK::DecorativeGeometry;
 /**
- * A class that holds the Display Attributes (Appearance) of an object displayed 
- * in the OpenSim Visualizer.
- * 
- * BaseAppearance contains Appearance properties that apply to all geometry.
- * Geometry that have a surface so that it can be textured can use the subclass
- * Appearance, while schematic line drawings (e.g. Arrows, Frames) can use 
- * LineAppearance which offers thickness. 
- *
- * TODO: Add Resolution to Appearance from DecorativeGeometry and 
- * utilize LineAppearance for Arrow, Line, and Frame (unused for now)
- *
- * @author Ayman Habib
- * @version 1.0
- */
-class OSIMSIMULATION_API BaseAppearance : public Object {
-    OpenSim_DECLARE_CONCRETE_OBJECT(BaseAppearance, Object);
+SurfaceAppearance class holds the display attributes  of a piece of Geometry 
+displayed in the OpenSim visualizer or GUI. The attributes in this 
+class are specific to geometry that have surfaces so these surfaces can be 
+textured, or rendered using  a variety of shading/lighting models etc. Attributes 
+common to all possible  objects that can be rendered (e.g. color, opacity, 
+visibility) are not included here but provided by the Appearance class instead.
+*/
+
+class OSIMSIMULATION_API SurfaceAppearance : public Object {
+    OpenSim_DECLARE_CONCRETE_OBJECT(SurfaceAppearance, Object);
 public:
     //==========================================================================
     // PROPERTIES
     //==========================================================================
-    OpenSim_DECLARE_PROPERTY(color, SimTK::Vec3,
-        "The color (rgb) used to display the object. ");
-    OpenSim_DECLARE_PROPERTY(opacity, double,
-        "The opacity (0-1) used to display the object. ");
     OpenSim_DECLARE_PROPERTY(representation, int,
-        "The representation (0:Hidden, 1:Points, 2:Wire 3:Shaded) used to display the object. ");
+        "The representation (1:Points, 2:Wire 3:Shaded) used to display the object. ");
+    OpenSim_DECLARE_OPTIONAL_PROPERTY(texture_file, std::string,
+        "Name of file containing texture. ");
 
     //--------------------------------------------------------------------------
     // CONSTRUCTION
     //--------------------------------------------------------------------------
 public:
-    BaseAppearance() {
-        constructProperties();
+    SurfaceAppearance() {
+        constructProperty_representation(3); // Surface shaded by default
     }
-    virtual ~BaseAppearance() {};
+    virtual ~SurfaceAppearance() {};
 
-private:
-    void constructProperties() {
-        constructProperty_color(SimTK::Vec3(1.0));
-        constructProperty_opacity(1.0);
-        constructProperty_representation(3);
+    bool hasTexture() {
+        return getProperty_texture_file().size() > 0;
```

Comment:
> Property also has the empty() method. @sherm1 referred to both in his comment; I'm not sure whether he meant std::string::empty() or Property<std::string>::empty().

I didn't notice that it was a Property rather than a string, but lucky for me Property has empty() also so I can look like I knew what I was saying!

>Would be nicer if optional properties had a "isPopulated" method to do this test.

Property objects are `(name, list-of-things)` pairs. They only differ in the allowable minimum and maximum lengths for the lists. "optional" is just a catchy name for a `(minLength=0,maxLength=1)` property list. So checking whether the list is `empty()` is the right thing to do.